### PR TITLE
Create a parser for ELEXON's FUELINST report.

### PR DESCRIPTION
**Pull request by unitrium**
## Issue

Report B1620 from ELEXON has gaps in the power generation at the moment and reports gas, nuclear production to 0. Our estimators are filling in the gaps for now but it is insufficient. The old report FUELINST still has the data but will be retired in 2023.

# Description

Temporarily use the FUELINST report in the ELEXON parser to fill in the missing UK data until the issue from the B1620 report is solved on ELEXON's side.

### CODE REVIEW

File: parsers/ELEXON.py [hunk 1]
```diff
@@ -50,6 +51,27 @@
     "Other": "unknown",
 }
 
+FUEL_INST_MAPPING = {
```
Comment (unitrium) [inline parsers/ELEXON.py:54]: The only thing is that this report doesn't give any indication on storage.

**Comment by pierresegonne** [comment]
Do we have any idea of what happens with the hydro storage? It does look like (looking at historical measured data) that there is discharge quite consistently every day

Screenshot 2022-12-20 at 09 01 50


Could we potentially get the discharge from the source we want to deprecate?

**Comment by unitrium** [comment]
Looking at the B1620, seems like hydro storage is also down, so we won't get any information from there until they solve the issue.
image

**Comment by unitrium** [comment]
By the way, we are only temporarily depreciating the B1620 as it is currently down. The FUELINST report is supposed to be retired in 2023, so we cannot rely in it forever.

**Comment by pierresegonne** [comment]
Do you have a link for getting that data? It is probably 0 most of the times except when discharging. 

Else yes let's just ignore it for now :/

Btw, I'm trying it out and when testing with a past datetime (ex: 2022-11-25), it does send the right request `https://api.bmreports.com/BMRS/FUELINST/v1?SettlementDate=2022-11-24&Period=%2A&ServiceType=csv&APIKey=xdz1cxjwbpjruec` but the data returned has as earliest settlement date 2022-12-19. 

Any idea why?

**Comment by unitrium** [comment]
For the B1620 report the link is :
https://api.bmreports.com/BMRS/B1620/v1?APIKey=YOURAPIKEY&ServiceType=csv&Period=*&SettlementDate=2022-12-19

**Comment by pierresegonne** [comment]
> For the B1620 report the link is 

I've looked at multiple days and it's all 0 indeed, so let's just roll with it, there is nothing we can do about it :/

**Comment by pierresegonne** [comment]
So besides this weird behaviour with past datetimes, LGTM.

**Comment by unitrium** [comment]
Actually it's a mistake of mine, we should use :
FromDateTime=<FromDateTime>&ToDateTime=<ToDateTime> parameters to query historical data.

**Comment by unitrium** [comment]
I will update the code to use this.

**Comment by mathilde-daugy** [comment]
image
There is something wrong with the values here, the orders of magnitude are still off.
Nuclear should be around 5 GW and wind around 12 GW not 50....
image

**Review by mathilde-daugy — APPROVED** [review]

File: parsers/ELEXON.py [hunk 2]
```diff
@@ -152,6 +176,68 @@ def parse_exchange(
...
+    df.rename(columns=mapping, inplace=True)
+    df["Settlement Date"] = df["Settlement Date"].apply(
+        lambda x: datetime.strptime(str(x), "%Y%m%d")
+    )
+    df["Settlement Period"] = df["Settlement Period"].astype(int)
+    df["datetime"] = df.apply(
+        lambda x: datetime_from_date_sp(x["Settlement Date"], x["Settlement Period"]),
+        axis=1,
+    )
+    data_points = list()
```
[the remark this answers is not in the document]
Comment (pierresegonne) [reply parsers/ELEXON.py:216]: does this fix the data issue? Let's not approve too hastily if it does not :)

[the remark this answers is not in the document]
Comment (unitrium) [reply parsers/ELEXON.py:227]: This was related to another issue that columns were not summed together correctly. Which got fixed, but I find @mathilde-daugy 's code better :)
There's no more data issue

**Pull request merged**